### PR TITLE
Make comparison op return bool

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR is stacked on https://github.com/NVIDIA/Fuser/pull/520

In C++, although we are allowed to overload comparison op to return an arbitrary type, it commonly just returns a `bool`. It is commonly assumed that comparison op will return `bool`, especially, in C++ standard library, such assumption also exist. For example, `std::vector<T>{} < std::vector<T>{}`, returns a `bool` instead of `decltype(T{} < T{})`. Due to this assumption, it actually makes more sense that we also return `bool`, instead of `DynamicType` for these ops.

Changing this means we will be unable to support `T`s whose `T < T` does not return a `bool` well, but in my entire career, I have never seen any of such case. The benefit of this switch is, we can now more seamlessly integrate with the oter parts of C++.

### CODE REVIEW

File: csrc/dynamic_type.h [hunk 1]
```diff
@@ -403,15 +403,135 @@ DEFINE_BINARY_OP(land, &&);
...
-DEFINE_BINARY_OP(eq, ==);
-DEFINE_BINARY_OP(neq, !=);
-DEFINE_BINARY_OP(lt, <);
-DEFINE_BINARY_OP(gt, >);
-DEFINE_BINARY_OP(le, <=);
-DEFINE_BINARY_OP(ge, >=);
 
 #undef DEFINE_BINARY_OP
 
+#define DEFINE_COMPARE_OP(opname, op)                                         \
```
Comment (zasdfgbnm) [inline csrc/dynamic_type.h:409]: Mostly copy-pasted from `DEFINE_BINARY_OP`, with modifications on return type handling.

File: csrc/dynamic_type.h [hunk 2]
```diff
@@ -403,15 +403,135 @@ DEFINE_BINARY_OP(land, &&);
...
+  /*TODO: we should inline the definition of lambdas into enable_if,*/        \
+  /*but I can only do this in C++20 */                                        \
+  constexpr auto opname##_defined_checker = [](auto x, auto y) constexpr {    \
+    using X = typename decltype(x)::type;                                     \
+    using Y = typename decltype(y)::type;                                     \
+    if constexpr (opcheck<X> op opcheck<Y>) {                                 \
+      return std::is_convertible_v<                                           \
```
Comment (zasdfgbnm) [inline csrc/dynamic_type.h:416]: Note that `std::is_convertible_v<From, To>`, is different from `opcheck<From>.canCastTo(opcheck<To>)`. The former is true only when `From` can be **implicitly** converted to `To`, while the latter is true also for explicit casts. For this case here, only allowing implicit casting makes more sense.

File: test/test_dynamic_type.cpp [hunk 3]
```diff
@@ -377,12 +377,48 @@ TEST_BINARY_OP_ALLTYPE(Mul, *);
...
 TEST_BINARY_OP_ALLTYPE(LogicalAnd, &&);
 TEST_BINARY_OP_ALLTYPE(LogicalOr, ||);
-TEST_BINARY_OP_ALLTYPE(Eq, ==);
-TEST_BINARY_OP_ALLTYPE(Ne, !=);
-TEST_BINARY_OP_ALLTYPE(Lt, <);
-TEST_BINARY_OP_ALLTYPE(Gt, >);
-TEST_BINARY_OP_ALLTYPE(Le, <=);
-TEST_BINARY_OP_ALLTYPE(Ge, >=);
+
+#define TEST_COMPARE_OP(name, op)                                              \
```
Comment (zasdfgbnm) [inline test/test_dynamic_type.cpp:381]: Mostly copy-pasted from `TEST_BINARY_OP_ALLTYPE` with modification.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**